### DISCUSSION TITLE
render surface normal map

### DISCUSSION THREAD

**Discussion by j-review** [mitsuba-renderer/mitsuba3#1230] [Q&A]
Dear Mitsuba community,

I am planning to render a surface normal map and I used _cbox_ example.
However, some surfaces show in black (I assume it is because it is in negative orientation)
Is there a way we could render surface normal for all surfaces?

Thanks in advance,
Guanzhou

![my_first_render8_11](https://github.com/user-attachments/assets/b2a23bed-f2cc-4c1b-92a6-2d107dc5bd1a)
![surface_normal](https://github.com/user-attachments/assets/0b40868e-ede1-4e8e-ac33-af9baa4fbf24)

Here is the complete XML file:

_<scene version="3.0.0">
    <default name="spp" value="256"/>
    <default name="res" value="1024"/>
    <default name="max_depth" value="6"/>

    <integrator type="aov">
        <string name="aovs" value="dd.y:uv,nn:geo_normal"/>
        <integrator type="path"/>
    </integrator>

    <sensor type="perspective" id="sensor">
        <string name="fov_axis" value="smaller"/>
        <float name="near_clip" value="0.001"/>
        <float name="far_clip" value="100.0"/>
        <float name="focus_distance" value="100"/>
        <float name="fov" value="120"/>
        <transform name="to_world">
            <lookat origin="0,  0,  2"
                    target="0,  0,  0"
                    up    ="0,  1,  0"/>
        </transform>
        <sampler type="independent">
            <integer name="sample_count" value="$spp"/>
        </sampler>
        <film type="hdrfilm">
            <integer name="width"  value="$res"/>
            <integer name="height" value="$res"/>
            <rfilter type="tent"/>
            <string name="pixel_format" value="rgb"/>
            <string name="component_format" value="float32"/>
        </film>
    </sensor>

    
    <bsdf type="diffuse" id="gray">
        <rgb name="reflectance" value="0.85, 0.85, 0.85"/>
    </bsdf>
    <bsdf type="diffuse" id="white">
        <rgb name="reflectance" value="0.885809, 0.698859, 0.666422"/>
    </bsdf>
    <bsdf type="diffuse" id="green">
        <rgb name="reflectance" value="0.105421, 0.37798, 0.076425"/>
    </bsdf>
    <bsdf type="diffuse" id="red">
        <rgb name="reflectance" value="0.570068, 0.0430135, 0.0443706"/>
    </bsdf>
    <bsdf type="dielectric" id="glass"/>
    <bsdf type="conductor" id="mirror"/>

    

    <shape type="obj" id="light">
        <string name="filename" value="meshes/cbox_luminaire.obj"/>
        <transform name="to_world">
            <translate x="0" y="-0.01" z="0"/>
        </transform>
        <ref id="white"/>
        <emitter type="area">
            <rgb name="radiance" value="18.387, 13.9873, 6.75357"/>
        </emitter>
    </shape>

    

    <shape type="obj" id="floor">
        <string name="filename" value="meshes/cbox_floor.obj"/>
        <ref id="white"/>
    </shape>

    <shape type="obj" id="ceiling">
        <string name="filename" value="meshes/cbox_ceiling.obj"/>
        <ref id="white"/>
    </shape>

    <shape type="obj" id="back">
        <string name="filename" value="meshes/cbox_back.obj"/>
        <ref id="white"/>
    </shape>

    <shape type="obj" id="greenwall">
        <string name="filename" value="meshes/cbox_greenwall.obj"/>
        <ref id="green"/>
    </shape>

    <shape type="obj" id="redwall">
        <string name="filename" value="meshes/cbox_redwall.obj"/>
        <ref id="red"/>
    </shape>

    <shape type="sphere" id="mirrorsphere">
        <transform name="to_world">
            <scale value="0.5"/>
            <translate x="-0.3" y="-0.5" z="0.2"/>
        </transform>
        <ref id="mirror"/>
    </shape>

    <shape type="sphere" id="glasssphere">
        <transform name="to_world">
            <scale value="0.25"/>
            <translate x="0.5" y="-0.75" z="-0.2"/>
        </transform>
        <ref id="glass"/>
    </shape>
</scene>_

**Comment by njroussel** [comment]
Hi @j-review 

The `aov` integrator will output the average normal vector within a pixel. The pixel values are therefore contained with [-1, -1, -1] and [1, 1, 1]. Indeed, your viewer does not know how to illustrate negative RGB values.

Here's a tiny python script to remap the image to values in  [0, 0, 0] [1, 1, 1]:

```python
import mitsuba as mi
mi.set_variant('scalar_rgb`)

bmp = mi.Bitmap("where/the/original/aov/image/is.exr")
tensor = mi.TensorXf(bmp)
tensor = (tensor + 1) / 2
mi.Bitmap(tensor).write("some/output/path.exr")
```

**Reply by merlinND** [reply]
@j-review An EXR image viewer like [tev](https://github.com/Tom94/tev/) will allow you to visualize the image directly without remapping, e.g. using the "FC" or "+/-" color modes
